File: Arrays/RotateArray/cpp/rotate_array.cpp
#include <iostream>
#include <cstdlib>
template<typename T>
void swap(T* x, T* y)
{
    T tmp = *x;
    *x = *y;
    *y = tmp;
}
/*
 * Rotate the given array with the size by the given rotate value
 * @param arr - the input array
 * @param size - the size of the array
 * @param rotate - the amount of times to rotate a number
 * @return none
 */
template<typename T>
void reverse(T* arr, std::size_t size)
{
    /* size = 3; stop at 1 */
    /* size = 4; stop at 2 */
    for(std::size_t i = size-1; i > (size/2)-1; i--)
    {
       /* 1 2 3 -> 3 2 1*/
       /* 8 7 4 2 -> 2 4 7 8 */ 
        swap(&arr[i], &arr[(size-1)-i]);
    } 
}

template<typename T>
void print_arr(T* arr, std::size_t size)
{
    for(std::size_t i = 0; i < size; i++)
    {
        std::cout<< arr[i] << ((i==size-1) ? "" : " ");
    }
    std::cout<<std::endl;
}

int main(int argc, char** argv)
{
    int hello[10] = {1,10,20,0,59,86,32,11,9,40};
    print_arr(hello,10);
    reverse(hello,10);
    print_arr(hello,10);
    return 0;
}

File: Arrays/RotateArray/java/RotateArray.java
import java.util.Arrays;

/*
 * Given an array of integers, rotate the array by N elements. Here, N is an integer.
 * Examples:
 * 1, 10, 20, 0, 59, 86, 32, 11, 9, 40
 * -1 rotations = 
 * 10, 20, 0 , 59, 86, 32, 11, 9, 40, 1
 * +2 rotations =
 * 9, 40, 1, 10, 20, 0, 59, 86, 32, 11 
 */
public class RotateArray {
    public static <T> void swap(T [] arr, int x, int y)
    {
        T tmp = arr[x];
        arr[x] = arr[y];
        arr[y] = tmp;
    }

    public <T> void reverse(T [] arr, int start, int end)
    {
        for(int i = end; i > ((end)/2); i--)
        {
            if(i > ((end-i)+start)) // this fixes the issue of swapping the middle twice
            {
                swap(arr, i, (end-i)+start); // calculate for negative indices
            }
        }
    }

    public <T> void rotateArray(T [] arr, int rotate)
    {
        rotate = rotate % arr.length;
        if(rotate<0)
        {
            rotate = rotate + arr.length;
        }
        this.reverse(arr,0,arr.length-1); // reverse the whole thing
        this.reverse(arr,0,rotate-1); // reverse the elements included in the rotation
        this.reverse(arr,rotate,arr.length-1); // reverse the upper half back to original order
    }

    public static void main(String [] args)
    {
        /* this has to be the Integer wrapper type because you can't convert from primitives to objects */
        Integer hello[] = {1,10,20,0,59,86,32,11,9,40};
        Integer trivial[] = {1,2,3};
        RotateArray r = new RotateArray();
        // positive rotation
        System.out.println(Arrays.toString(hello));
        r.rotateArray(hello, 2);
        System.out.println(Arrays.toString(hello));
        // negative rotation
        System.out.println(Arrays.toString(trivial));   
        r.rotateArray(trivial, -1);
        System.out.println(Arrays.toString(trivial));
        
    }
}
